test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -109,310 +112,334 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
-		}
+		r.Vs, err = dc.Resource(SnapshotGVR).Namespace(r.Vs.GetNamespace()).Get(context.TODO(), r.Vs.GetName(), metav1.GetOptions{})
+
+		snapshotStatus := r.Vs.Object["status"].(map[string]interface{})
+		snapshotContentName := snapshotStatus["boundVolumeSnapshotContentName"].(string)
+		framework.Logf("received snapshotStatus %v", snapshotStatus)
+		framework.Logf("snapshotContentName %s", snapshotContentName)
+		framework.ExpectNoError(err)
 
-		ginkgo.By("checking the SnapshotContent has been deleted")
-		err = utils.WaitForGVRDeletion(l.dc, SnapshotContentGVR, snapshotContentName, framework.Poll, framework.SnapshotDeleteTimeout)
+		r.Vscontent, err = dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
 		framework.ExpectNoError(err)
-	} else {
-		framework.Failf("Invalid test config. DeletionPolicy should be either Delete or Retain. DeletionPolicy: %v", sct.DeletionPolicy)
+	case testpatterns.PreprovisionedCreatedSnapshot:
+		// prepare a pre-provisioned VolumeSnapshotContent with certain data
+		// Because this could be run with an external CSI driver, we have no way
+		// to pre-provision the snapshot as we normally would using their API.
+		// We instead dynamically take a snapshot and create another snapshot using
+		// the first snapshot's snapshot handle.
+		ginkgo.Skip("Preprovisioned test not implemented")
+		ginkgo.By("taking a snapshot with deletion policy retain")
+		ginkgo.By("recording the volume handle and status.snapshotHandle")
+		ginkgo.By("deleting the snapshot and snapshot content") // TODO: test what happens when I have two snapshot content that refer to the same content
+		ginkgo.By("creating a snapshot content with the snapshot handle")
+		ginkgo.By("creating a snapshot with that snapshot content")
 	}
+	return &r
 }
 
-// WaitForSnapshotReady waits for a VolumeSnapshot to be ready to use or until timeout occurs, whichever comes first.
-func WaitForSnapshotReady(c dynamic.Interface, ns string, snapshotName string, poll, timeout time.Duration) error {
-	framework.Logf("Waiting up to %v for VolumeSnapshot %s to become ready", timeout, snapshotName)
-
-	if successful := utils.WaitUntil(poll, timeout, func() bool {
-		snapshot, err := c.Resource(SnapshotGVR).Namespace(ns).Get(context.TODO(), snapshotName, metav1.GetOptions{})
-		if err != nil {
-			framework.Logf("Failed to get snapshot %q, retrying in %v. Error: %v", snapshotName, poll, err)
-			return false
+// CleanupResource cleans up the snapshot resource and ignores not found errors
+func (sr *SnapshotResource) CleanupResource() error {
+	var err error
+	var cleanupErrs []error
+
+	dc := sr.Config.Framework.DynamicClient
+
+	if sr.Vs != nil {
+		framework.Logf("deleting snapshot %q/%q", sr.Vs.GetNamespace(), sr.Vs.GetName())
+
+		sr.Vs, err = dc.Resource(SnapshotGVR).Namespace(sr.Vs.GetNamespace()).Get(context.TODO(), sr.Vs.GetName(), metav1.GetOptions{})
```

Comment:
I think this CleanupResource() organization is a big win. But I want to make sure I understand what's going on.

It looks like the semantics have changed here: if a VolumeSnapshot isn't found, the previous test would fail, whereas this one would continue (see https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L335 I think).

Is that because this may not have been called after WaitForSnapshotReady? Or another reason?

Reply:
The big idea with CleanupResource is that it's purpose should be to clean up as much as it can to prevent resource leaks and be resilient to weird conditions, instead of acting as a way to test the deletion works.

The semantics of deletion should be captured by the deletion test: https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go#L205. If the semantics of the test there has changed that would be an issue. If the volume snapshot was not found in that test then https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go#L281 would throw an error.

Reply:
but line 281 won't fire if the resource is not found---don't we want that to be a test error?

+1 on making CleanupResource robust and doing the testing in the test body, that's very nice organization.

Reply:
I suppose you're right. Although CreateSnapshotResource does wait for the snapshot to be ready before it returns, and will error if it doesnt get created. https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go#L332

I suppose implicitly relying on the creation step to guarantee the snapshot exists prior to deletion would be confusing. It may just be easier for the DeleteAndWaitSnapshot method to not ignore `not found` errors.

Reply:
I changed the behavior of DeleteAndWaitSnapshot to not ignore not found errors, and to return the error instead of failing. The test now checks that the deletion succeeds, so it should catch any not found errors.